test/integration/whois_server_integration_test.rb

```diff
@@ -0,0 +1,225 @@
+require 'test_helper'
+require 'simpleidn'
+require_relative '../../lib/whois_server'
+require_relative '../../lib/whois_server_core'
+require_relative '../../app/models/whois_record'
+require_relative '../../app/validators/unicode_validator'
+
+class WhoisServerIntegrationTest < Minitest::Test
+  include WhoisServer
+
+  def setup
+    super
+    @sent_data = []
+    @connection_closed = false
+    @connection_closed_after_writing = false
+    @logger_output = StringIO.new
+    @logger = Logger.new(@logger_output)
+    ENV['WHOIS_ENV'] = 'test'
+  end
+
+  def test_receive_data_with_invalid_data
+    stubs(:invalid_data?).returns(true)
+    stubs(:invalid_encoding_msg).returns('Invalid encoding')
+    stubs(:connection).returns(nil)
+
+    receive_data('invalid\xFF')
+
+    assert_equal 1, @sent_data.length
+    assert_includes @sent_data.first, 'Invalid encoding'
+    assert @connection_closed_after_writing
+  end
+
+  def test_receive_data_with_no_ip
+    stubs(:extract_ip).returns(nil)
+    stubs(:connection).returns(nil)
+
+    receive_data('example.ee')
+
+    assert_equal 0, @sent_data.length
+  end
+
+  def test_receive_data_full_flow
+    data = 'example.ee'
+    ip = ['12345', '127.0.0.1']
+    mock_record = Minitest::Mock.new
+    mock_record.expect(:unix_body, 'Whois body')
+    mock_record.expect(:id, 1)
+
+    stubs(:connection).returns(nil)
+    stubs(:extract_ip).returns(ip)
+    stubs(:invalid_data?).returns(false)
+    WhoisRecord.stubs(:find_by).returns(mock_record)
+
+    receive_data(data)
+
+    assert_equal 1, @sent_data.length
+    assert @connection_closed_after_writing
+    mock_record.verify
+  end
+
+  def test_receive_data_integration_full_flow
+    data = 'integration-test.ee'
+    ip = ['12345', '127.0.0.1']
+    
+    stubs(:connection).returns(nil)
+    stubs(:extract_ip).returns(ip)
+    stubs(:invalid_data?).returns(false)
+    
+    mock_record = Minitest::Mock.new
+    mock_record.expect(:unix_body, 'Integration test body')
+    mock_record.expect(:id, 1)
+    
+    WhoisRecord.stubs(:find_by).returns(mock_record)
+    
+    receive_data(data)
+    
+    assert_equal 1, @sent_data.length
+    assert @connection_closed_after_writing
+    mock_record.verify
+  end
+
+  def test_process_whois_request_with_invalid_encoding
+    data = "\xFF\xFE".force_encoding('ASCII-8BIT')
+    ip = ['12345', '127.0.0.1']
+    
+    send(:process_whois_request, data, ip)
+
+    assert_equal 1, @sent_data.length
+    assert_includes @sent_data.first, 'invalid encoding'
+  end
+
+  def test_process_whois_request_with_invalid_domain_format
+    data = 'invalid..domain.ee'
+    ip = ['12345', '127.0.0.1']
+
+    send(:process_whois_request, data, ip)

```

Comment:
Same thing here. The issue was that when we ran code like `printf 'täpiline.ee\r\n' | nc localhost 43`, the WHOIS service would crash. That’s why we updated the `process_whois_request` method to add extra checks: specifically for cases when the `invalid_data?` method in the public scope doesn’t see anything suspicious.

Honestly, I’d slightly restructure and refactor this, but the problem is that if I move some of the logic from private methods up to a higher-level abstraction, for example, from `process_whois_request` into `receive_data`, the tests will start failing. And that’s exactly the point I’m making: you shouldn’t be testing private methods directly. Public methods should be the ones calling them as needed, depending on the data and conditions.